### PR TITLE
Fix overlapping domains in Dialyzer for IO module

### PR DESCRIPTION
**Pull request by eksperimental**
Originally disussed in https://github.com/elixir-lang/elixir/pull/9046#discussion_r284604298

### CODE REVIEW

File: lib/elixir/lib/io.ex [hunk 1]
```diff
@@ -415,8 +415,9 @@ defmodule IO do
   See `IO.getn/3` for a description of return values.
 
   """
-  @spec getn(chardata | String.Chars.t(), pos_integer) :: chardata | nodata
-  @spec getn(device, chardata | String.Chars.t()) :: chardata | nodata
+  @spec getn(prompt | device, count | prompt) :: chardata | nodata
+        when prompt: chardata | String.Chars.t(),
+             count: pos_integer
```
Comment (josevalim) [inline lib/elixir/lib/io.ex]: The problem with this typespec is that it implies that `chardata, prompt` will work, but it won't. Given typespecs are not used by Dialyzer, they primary purpose is docs, and I think the docs in this case would then become misleading.

Having to choose between both, I would prefer to keep the docs bit accurate.

Comment (eksperimental) [reply lib/elixir/lib/io.ex]: Why do you say typespecs are not used by dialyzer?

Comment (ericmj) [reply lib/elixir/lib/io.ex]: Dialyzer can verify the typespecs against its own analysis, but they are not used for the analysis itself.

**Comment by josevalim** [comment]
Closing this as we don't plan to move it forward, thanks @eksperimental !

**Pull request closed without merging**